File: benchmark/external/rinla/iid_poisson/iid_poisson_compare.R
# Fit toy IID Poisson with R-INLA and dump the per-x marginals + τ
# marginal for cross-validation against Latte's INLA.
#
# Model: y_i ~ Poisson(exp(x_i)) with x_i ~ N(0, 1/τ) iid,
#        τ ~ PCPrior.Precision(1, α=0.01).
#
# Inputs:
#   data.csv      — y, idx
#   params.json   — { "pc_U", "pc_alpha", "strategy" }
#
# Outputs:
#   rinla_x_marginals.csv  — long form (i, x, density)
#   rinla_tau_marginal.csv — (x, density)
#   rinla_x_summary.csv    — per-i mean / sd / quantiles
#   rinla_meta.json

suppressPackageStartupMessages({
    library(INLA)
    library(jsonlite)
})

argv <- commandArgs(trailingOnly = TRUE)
input_dir <- if (length(argv) >= 1) argv[1] else "."
output_dir <- if (length(argv) >= 2) argv[2] else input_dir

data <- read.csv(file.path(input_dir, "data.csv"))
params <- fromJSON(file.path(input_dir, "params.json"))
strategy <- params$strategy
n <- nrow(data)

cat(sprintf("R-INLA IID Poisson: n = %d, strategy = %s\n", n, strategy))

formula <- y ~ -1 + f(idx,
    model = "iid",
    hyper = list(prec = list(prior = "pc.prec",
                             param = c(params$pc_U, params$pc_alpha)))
)

result <- inla(formula,
    data = data,
    family = "poisson",
    control.inla = list(strategy = strategy, int.strategy = "grid"),
    control.compute = list(return.marginals.predictor = TRUE),
)

x_long <- do.call(rbind, lapply(seq_len(n), function(i) {
    m <- result$marginals.random$idx[[i]]
    data.frame(i = i, x = m[, "x"], density = m[, "y"])
}))
write.csv(x_long, file.path(output_dir, "rinla_x_marginals.csv"), row.names = FALSE)

tau_marg <- result$marginals.hyperpar$`Precision for idx`
tau_df <- data.frame(x = tau_marg[, "x"], density = tau_marg[, "y"])
write.csv(tau_df, file.path(output_dir, "rinla_tau_marginal.csv"), row.names = FALSE)

x_summary <- do.call(rbind, lapply(seq_len(n), function(i) {
    s <- result$summary.random$idx[i, ]
    data.frame(
        name = sprintf("x[%d]", i),
        mean = s$mean, sd = s$sd,
        q025 = s$`0.025quant`, q5 = s$`0.5quant`, q975 = s$`0.975quant`
    )
}))
write.csv(x_summary, file.path(output_dir, "rinla_x_summary.csv"), row.names = FALSE)

meta <- list(
    strategy = strategy, n = n,
    inla_version = as.character(packageVersion("INLA")),
    elapsed_seconds = result$cpu.used[["Total"]],
    status = "ok"
)
writeLines(toJSON(meta, auto_unbox = TRUE, pretty = TRUE),
    file.path(output_dir, "rinla_meta.json"))

cat("done.\n")

File: benchmark/external/rinla/iid_poisson/iid_poisson_compare.jl
# Toy IID Poisson: Latte INLA vs R-INLA, both with `simplified.laplace`.
#
# Synthetic n=50 dataset, IID Gaussian random effects, Poisson observations.
# A simpler-than-Tokyo benchmark: no smoothing prior, no per-day correlation.
# Per-x marginals are KS-compared against R-INLA's output.

using Pkg
Pkg.activate(joinpath(@__DIR__, "..", "..", ".."))

using CSV
using DataFrames
using Distributions
using DynamicPPL          # full module: the @latte macro's expansion references it
using GaussianMarkovRandomFields
using GaussianMarkovRandomFields: IIDModel
using JSON3
using Latte
using Printf
using StableRNGs
using Statistics

const WORKDIR = joinpath(@__DIR__, "_workdir")
const N = 50
const SEED = UInt64(0x0badcafe)

@latte function iid_poisson_model(y, n)
    τ ~ PCPrior.Precision(1.0, α = 0.01)
    x ~ IIDModel(n)(τ = τ)
    for i in eachindex(y)
        y[i] ~ Poisson(exp(x[i]); check_args = false)
    end
end

function generate_data(n::Int, seed::UInt64)
    rng = StableRNG(seed)
    true_x = randn(rng, n) .* 0.4 .+ 0.7
    y = rand.(rng, Poisson.(exp.(true_x)))
    return (; n = n, y = y, true_x = true_x)
end

function _user_x_marginals(result, sym::Symbol = :x)
    groups = Latte.latent_groups(result)
    return [result.latent_marginals[i] for i in groups[sym]]
end

function _ks_density(
        engine, ref_grid::Vector{Float64}, ref_density::Vector{Float64},
    )
    n = length(ref_grid)
    cdf_vals = Vector{Float64}(undef, n)
    cdf_vals[1] = 0.0
    for k in 2:n
        cdf_vals[k] = cdf_vals[k - 1] +
            0.5 * (ref_density[k - 1] + ref_density[k]) * (ref_grid[k] - ref_grid[k - 1])
    end
    Z = cdf_vals[end]
    Z > 0 && (cdf_vals ./= Z)
    best_abs = 0.0
    best_signed = 0.0
    for k in 1:n
        gap = cdf(engine, ref_grid[k]) - cdf_vals[k]
        if abs(gap) > best_abs
            best_abs = abs(gap)
            best_signed = gap
        end
    end
    return best_abs, best_signed
end

function main(args::Vector{String} = ARGS)
    mkpath(WORKDIR)
    data = generate_data(N, SEED)
    @info "IID Poisson dataset" n = data.n sum_y = sum(data.y) mean_y = mean(data.y)

    # Default: the @latte macro builds a compact LGM and inla resolves the VBC
    # mean correction. `--augmented` opts into the legacy augmented + SLA mode.
    augmented = "--augmented" in args
    marg = augmented ? SimplifiedLaplace() : nothing   # nothing ⇒ resolve (→ VBC, compact LTM)
    @info "running Latte INLA" mode = (augmented ? "augmented + simplified.laplace (legacy)" : "compact + VBC (default)")
    lgm = iid_poisson_model(data.y, data.n; augment = augmented)
    t_latte = @elapsed result = inla(
        lgm, data.y;
        latent_marginalization_method = marg,
        progress = false,
    )
    @info "Latte done" elapsed = round(t_latte, digits = 2) augmented resolved = string(typeof(augmented ? SimplifiedLaplace() : Latte.default_marginalization(lgm)).name.name)
    latte_x = _user_x_marginals(result)

    rinla_marker = joinpath(WORKDIR, "rinla_x_marginals.csv")
    if !isfile(rinla_marker) || "--refresh-rinla" in args
        @info "running R-INLA"
        df = DataFrame(idx = 1:data.n, y = data.y)
        CSV.write(joinpath(WORKDIR, "data.csv"), df)
        open(joinpath(WORKDIR, "params.json"), "w") do io
            JSON3.write(
                io, Dict(
                    "pc_U" => 1.0, "pc_alpha" => 0.01,
                    "strategy" => "simplified.laplace",
                )
            )
        end
        rscript = joinpath(@__DIR__, "iid_poisson_compare.R")
        t_rinla = @elapsed run(`Rscript $rscript $(WORKDIR) $(WORKDIR)`)
        @info "R-INLA done" elapsed = round(t_rinla, digits = 2)
    else
        @info "loading R-INLA cache" workdir = WORKDIR
    end

    rinla_meta = JSON3.read(read(joinpath(WORKDIR, "rinla_meta.json"), String))
    rinla_marg_df = CSV.read(joinpath(WORKDIR, "rinla_x_marginals.csv"), DataFrame)

    ks_per_i = Float64[]
    sign_per_i = Float64[]
    for i in 1:data.n
        sub = filter(row -> row.i == i, rinla_marg_df)
        sort!(sub, :x)
        ks, sgn = _ks_density(
            latte_x[i], Vector{Float64}(sub.x), Vector{Float64}(sub.density),
        )
        push!(ks_per_i, ks)
        push!(sign_per_i, sgn)
    end

    worst_i = argmax(ks_per_i)
    println()
    println("Toy IID Poisson — Latte vs R-INLA (simplified.laplace), n=$(data.n)")
    println("="^70)
    @printf "%-30s max %.4f (i*=%d)   median %.4f   count > 0.05: %d / %d\n" "x_i KS" ks_per_i[worst_i] worst_i sort(ks_per_i)[div(length(ks_per_i), 2) + 1] count(>(0.05), ks_per_i) data.n
    @printf "%-30s signed at argmax: %+.4f\n" "" sign_per_i[worst_i]
    @printf "%-30s INLA version: %s,  elapsed: %.2f s (Latte) vs %.2f s (R-INLA)\n" "" String(rinla_meta.inla_version) t_latte Float64(rinla_meta.elapsed_seconds)
    println()

    return (ks_per_i = ks_per_i, sign_per_i = sign_per_i, worst_i = worst_i)
end

if abspath(PROGRAM_FILE) == @__FILE__
    main()
end
